service/history/workflow/workflow_task_state_machine.go

```diff
@@ -1087,11 +1080,64 @@ func (m *workflowTaskStateMachine) afterAddWorkflowTaskCompletedEvent(
 	attrs := event.GetWorkflowTaskCompletedEventAttributes()
 	m.ms.executionInfo.LastCompletedWorkflowTaskStartedEventId = attrs.GetStartedEventId()
 	m.ms.executionInfo.MostRecentWorkerVersionStamp = attrs.GetWorkerVersion()
-	if err := m.ms.CompleteDeploymentTransition(attrs.GetVersioningBehavior()); err != nil {
-		return err
+
+	wftDeployment := attrs.GetDeployment()
+	wftBehavior := attrs.GetVersioningBehavior()
+	versioningInfo := m.ms.GetExecutionInfo().GetVersioningInfo()
+
+	var completedTransition bool
+	if versioningInfo.DeploymentTransition != nil {
+		// It's possible that the completed WFT is not yet from the current transition because when
+		// the transition started, the current wft was already started. In this case, we allow the
+		// started wft to run and when completed, we create another wft immediately.
+		if versioningInfo.DeploymentTransition.GetDeployment().Equal(wftDeployment) {
+			versioningInfo.DeploymentTransition = nil
+			completedTransition = true
+		}
+	}
+
+	if m.ms.GetDeploymentTransition() != nil {
+		// There is still a transition going on. We need to schedule a new WFT so it goes to the
+		// transition deployment this time.
+		if _, err := m.ms.AddWorkflowTaskScheduledEvent(
+			false,
+			enumsspb.WORKFLOW_TASK_TYPE_NORMAL,
+		); err != nil {
+			return err
+		}
+	} else {
+		// Deployment and behavior before applying the data came from the completed wft.
+		wfDeploymentBefore := m.ms.GetEffectiveDeployment()
+		wfBehaviorBefore := m.ms.GetEffectiveVersioningBehavior()
```

Comment:
It really feels like we need a message/struct which combines `Deployment` and `Behavior`.

Reply:
I looked at the usages. Whenever Behavior is used, Deployment is also used, but not vice versa (in many places we only care about deployment). I think we can keep it as is.